lapack/gonum/dpbtrf.go

```diff
@@ -0,0 +1,188 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+func (impl Implementation) Dpbtrf(uplo blas.Uplo, n, kd int, ab []float64, ldab int) (ok bool) {
+	const nbmax = 32
+
+	switch {
+	case uplo != blas.Upper && uplo != blas.Lower:
+		panic(badUplo)
+	case n < 0:
+		panic(nLT0)
+	case kd < 0:
+		panic(kdLT0)
+	case ldab < kd+1:
+		panic(badLdA)
+	}
+
+	// Quick return if possible.
+	if n == 0 {
+		return true
+	}
+
+	if len(ab) < (n-1)*ldab+kd {
+		panic(shortAB)
+	}
+
+	opts := string(blas.Upper)
+	if uplo == blas.Lower {
+		opts = string(blas.Lower)
+	}
+	nb := impl.Ilaenv(1, "DPBTRF", opts, n, kd, -1, -1)
+	// The block size must not exceed the semi-bandwidth kd, and must not
+	// exceed the limit set by the size of the local array work.
+	nb = min(nb, nbmax)
+
+	if nb <= 1 || kd < nb {
+		// Use unblocked code.
+		return impl.Dpbtf2(uplo, n, kd, ab, ldab)
+	}
+
+	// Use blocked code.
+	ldwork := nb
+	work := make([]float64, nb*ldwork)
+	bi := blas64.Implementation()
+	if uplo == blas.Upper {
+		// Compute the Cholesky factorization of a symmetric band
+		// matrix, given the upper triangle of the matrix in band
+		// storage.
+
+		// Process the band matrix one diagonal block at a time.
+		for i := 0; i < n; i += nb {
+			ib := min(nb, n-i)
+			// Factorize the diagonal block.
+			ok := impl.Dpotf2(uplo, ib, ab[i*ldab:], ldab-1)
+			if !ok {
+				return false
+
+			}
+			if i+ib >= n {
```

Comment:
`>`? The Fortran has `i+ib.LE.n`. Is this a 1 versus 0?

Reply:
Yes, 1 vs 0 and the inverted condition. Another way to see it, `i+ib` is the _index_ where the next block would begin. Maybe it would be clearer to break out of the loop here and remove the end test in the for loop above. I just didn't want to have two apparent places to end the loop.